Added CLI command structure

matthanley opened the pull request:
Added CLI commands for REPL and timetravel functionality

- Added TimeTravel command for updating test data with a date offset
- Added REPL command for access to models and knex in development
- Added pattern for creating new CLI commands, including
  - User input
  - Output
  - Validation of `NODE_ENV`
- TimeTravel command is in the main Ghost repo because it requires the schema

Reference material: https://laravel.com/docs/9.x/artisan

matthanley committed:
Implemented CLI command structure

matthanley committed:
Added confirmation

matthanley committed:
Cleanup

matthanley committed:
Switch from prompt-sync to inquirer

matthanley committed:
Removed tinker as an alias for repl

matthanley committed:
Added progress bar

daniellockyer commented on package.json:
```diff
@@ -194,6 +194,7 @@
     "@playwright/test": "1.22.0",
     "@tryghost/express-test": "0.10.0",
     "c8": "7.11.2",
+    "cli-progress": "^3.11.0",
```
Can you pin these dependencies?

daniellockyer commented on core/cli/timetravel.js:
```diff
@@ -0,0 +1,62 @@
+            .max(datum.column, {as: datum.column})
+            .first();
+
+        const dateOffset = Math.floor(
```
We should probably handle `members_login_events` being empty, otherwise I get:

```
...
[2022-05-16 17:32:48] INFO Timetravel will use an offset of +NaN days
...
```

daniellockyer commented on core/cli/repl.js:
```diff
@@ -0,0 +1,21 @@
+    async handle() {
+        const repl = require('repl');
+        const cli = repl.start('> ');
+        this.initializeContext(cli.context);
```
Can we add maybe a one-liner detailing the variables available?

matthanley committed:
Updated base Command class to be more flexible

- Switched from @tryghost/logging to our CLI package pretty-cli
- Dropped the use of a Proxy for a simpler implementation
- Added argument parsing functionality
- Added a check and a fallback for TimeTravel when member logins don't exist
- Added help screens
- Added output hints to REPL command
- Pinned dependencies

matthanley committed:
Merge branch 'main' into cli-commands

daniellockyer commented on core/cli/command.js:
```diff
@@ -0,0 +1,122 @@
+const cli = require('@tryghost/pretty-cli');
```
I think this dependency is hoisted from gscan's dependencies and we need to add it to package.json here too

daniellockyer commented on core/cli/timetravel.js:
```diff
@@ -0,0 +1,87 @@
+const Command = require('./command');
+const chalk = require('chalk');
```
Same with chalk 🤔

matthanley committed:
Added and pinned dependencies
